### PR TITLE
Add script and job to automate Go updates

### PR DESCRIPTION
**Pull request by simonpasquier**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -35,3 +42,13 @@ workflows:
         filters:
           branches:
             only: master
+  nightly:
+    triggers:
+      - schedule:
+          cron: "0 0 * * *"
```
Comment (SuperQ) [inline .circleci/config.yml]: Let's randomize this value a bit so we avoid running everything at midnight. :smile:

File: scripts/bump_version.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+if [ -z "${GITHUB_TOKEN}" ]; then
+	echo -e "\e[31mGitHub token (GITHUB_TOKEN) not set. Terminating.\e[0m"
+	exit 1
+fi
+
+go run ./cmd/builder-bumper
+
+if [ -n "$(git status --porcelain)" ]; then
+    git checkout -b bump_version
+	git config user.email "${GITHUB_MAIL}"
```
Comment (SuperQ) [inline scripts/bump_version.sh]: Looks like inconsistent use of tabs and spaces for indeting.

I prefer we stick to two-space indent for shell code.

File: scripts/bump_version.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+# GITHUB_TOKEN required scope: repo.repo_public
+GITHUB_TOKEN="${GITHUB_TOKEN:-}"
+if [ -z "${GITHUB_TOKEN}" ]; then
+	echo -e "\e[31mGitHub token (GITHUB_TOKEN) not set. Terminating.\e[0m"
+	exit 1
+fi
+
+go run ./cmd/builder-bumper
+
```
Comment (SuperQ) [inline scripts/bump_version.sh]: Let's check `$?` for failed go run.

File: scripts/bump_version.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+# Setting -x is absolutely forbidden as it could leak the GitHub token.
+set -uo pipefail
+
+
+GITHUB_MAIL="${GITHUB_MAIL:-prometheus-team@googlegroups.com}"
+GITHUB_USER="${GITHUB_USER:-prombot}"
+
+# GITHUB_TOKEN required scope: repo.repo_public
+GITHUB_TOKEN="${GITHUB_TOKEN:-}"
+if [ -z "${GITHUB_TOKEN}" ]; then
```
Comment (SuperQ) [inline scripts/bump_version.sh]: Minor bash nit, use `[[ ]]` for bash tests.

File: .circleci/config.yml (hunk 1 above)
Comment (simonpasquier) [reply .circleci/config.yml]: Bumped to 6am as Go releases happen at PST times usually.

File: scripts/bump_version.sh (hunk 2 above)
Comment (simonpasquier) [reply scripts/bump_version.sh]: Done

File: scripts/bump_version.sh (hunk 3 above)
Comment (simonpasquier) [reply scripts/bump_version.sh]: Added `set -e` at the beginning of the script.

File: scripts/bump_version.sh (hunk 4 above)
Comment (simonpasquier) [reply scripts/bump_version.sh]: done

File: scripts/bump_version.sh (hunk 3 above)
Comment (SuperQ) [reply scripts/bump_version.sh]: I prefer to avoid `set -e` as it causes failures to be silent.

**Comment by simonpasquier** [comment]
@SuperQ it should be ready for another review.

File: scripts/bump_version.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# GITHUB_TOKEN required scope: repo.repo_public
+GITHUB_TOKEN="${GITHUB_TOKEN:-}"
+if [[ -z ${GITHUB_TOKEN} ]]; then
+  echo -e "\e[31mGitHub token (GITHUB_TOKEN) not set. Terminating.\e[0m"
+  exit 1
+fi
+
+go run ./cmd/builder-bumper
+
+if [[ -n $(git status --porcelain) ]]; then
```
Comment (SuperQ) [inline scripts/bump_version.sh]: I would quote the subshell. `"$()"`.

File: scripts/bump_version.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  git config user.email "${GITHUB_MAIL}"
+  git config user.name "${GITHUB_USER}"
+  git add .
+  git commit -s -m "Bump Go version"
+  # stdout and stderr are redirected to /dev/null otherwise git-push could leak the token in the logs.
+  if git push --quiet "https://${GITHUB_TOKEN}:@github.com/prometheus/golang-builder" --set-upstream bump_version 1>/dev/null 2>&1; then
+    curl --show-error --silent \
+      -u "${GITHUB_USER}:${GITHUB_TOKEN}" \
+      -X POST \
+      -d "{\"title\":\"Bump Go version\",\"base\":\"master\",\"head\":\"bump_version\",\"body\":\"\"}" \
```
Comment (SuperQ) [inline scripts/bump_version.sh]: Maybe use single quotes here to avoid lots of `\`.

**Comment by simonpasquier** [comment]
@SuperQ it should be ok now.

File: scripts/bump_version.sh (hunk 6 above)
Comment (SuperQ) [reply scripts/bump_version.sh]: Why not this:

```bash
-d '{"title":"Bump Go version","base":"master","head":"bump_version","body":""}'
```

Comment (simonpasquier) [reply scripts/bump_version.sh]: Because I probably lack some coffee when I pushed this ;-)

Comment (simonpasquier) [reply scripts/bump_version.sh]: In fact the change was still on my local machine and never made it to GitHub

Comment (SuperQ) [reply scripts/bump_version.sh]: :joy:

**Comment by simonpasquier** [comment]
The first run of the script failed because the Circle CI's project doesn't have the GITHUB_TOKEN configured. I've sent #68 to use org context instead and after it is merged, we would just need to add the token to the org context.

**Pull request merged**